### PR TITLE
PAO virtual orbitals

### PR DESCRIPTION
**Pull request by henankf223**
## Description
This PR add the option to use Projected Atomic Orbitals (PAOs) as virtual orbitals in embedding computations.

## User Notes
- [x] Features added: PAO
- [x] Changes to compilation: src/orbital-helpers/pao_builder
- [x] Fix test case embedding-3
- [x] Simplify some of the embedding code (src/orbital-helpers/orbital_embedding)

## Checklist
- [x] Added tests of new features
- [x] Removed comments in code and input files
- [x] Documented source code
- [x] Checked for redundant headers
- [ ] Checked for consistency in the formatting of the output file
- [ ] Documented new features in the manual
- [ ] Ready to go!

### CODE REVIEW

File: src/orbital-helpers/orbital_embedding.cc [hunk 1]
```diff
@@ -1,689 +1,772 @@
...
+                if (lv->get(0, i) > 0.5) {
+                    outfile->Printf("\n Warning! Virtual orbital %d have eigenvalue (%8.8f) "
+                                    "larger than 0.5 are partitioned to B!",
+                                    i + offset_vec, lv->get(0, i));
+                }
+            }
+        }
+    }
+
+    if (options.get_str("EMBEDDING_VIRTUAL_SPACE") == "PAO") {
```
Comment (henankf223) [inline src/orbital-helpers/orbital_embedding.cc]: This section is what I added to the old code

File: src/orbital-helpers/orbital_embedding.cc [hunk 2]
```diff
@@ -1,689 +1,772 @@
...
+                index_A_vir.push_back(i + offset_vec);
+                outfile->Printf("\n Orbital %d built as PAO", i + offset_vec + 1);
+            } else {
+                index_B_vir.push_back(i + offset_vec);
+            }
+        }
+
+        outfile->Printf("\n ****** PAOs done ******");
+    }
+
```
Comment (henankf223) [inline src/orbital-helpers/orbital_embedding.cc]: After building PAOs, the code will run normally, with only virtual orbital changes to PAO

File: pymodule.py [hunk 3]
```diff
@@ -304,8 +304,8 @@ def orbital_projection(ref_wfn, options, mo_space_info):
     # Create the fragment(embedding) projector and apply to rotate orbitals
     if options.get_bool("EMBEDDING"):
         forte.print_method_banner(["Frozen-orbital Embedding", "Nan He"]);
-        pf = forte.make_fragment_projector(ref_wfn, options)
-        return forte.make_embedding(ref_wfn, options, pf, mo_space_info)
+        pf, nbfA = forte.make_fragment_projector(ref_wfn, options)
+        return forte.make_embedding(ref_wfn, options, pf, nbfA, mo_space_info)
```
Comment (fevangelista) [inline pymodule.py]: I think that renaming variables would be helpful
- `pf` -> `fragment_projector`
- `nbfA` -> `fragment_nbf`

File: register_forte_options.py [hunk 4]
```diff
@@ -158,6 +158,15 @@ def register_embedding_options(forte_options):
     forte_options.add_int(
         "EMBEDDING_ADJUST_B_UOCC", 0,
         "Adjust number of virtual orbitals between A and B, +: move to B, -: move to A")
+    forte_options.add_str(
+        "EMBEDDING_VIRTUAL_SPACE", "DIRECT", 
+        "Vitual space scheme, DIRECT, PAO or IAO.")
```
Comment (fevangelista) [inline register_forte_options.py]: When there is only a finite number of valid options, it is better to use the other versions of `forte_options.add_str` that allows to specify the allowed values. Basically, just add `['DIRECT', 'PAO', 'IAO'] after ` "DIRECT",`. Also, what does DIRECT stand for? It sounds very generic, what about DMET or ASET?

File: src/orbital-helpers/fragment_projector.h [hunk 5]
```diff
@@ -56,8 +56,16 @@ class FragmentProjector {
...
-                                        std::shared_ptr<psi::BasisSet> basis);
+	psi::SharedMatrix build_f_projector(std::shared_ptr<psi::BasisSet> basis);
+
+    /// Return number of basis functions in the system
+    int get_nbf_A() { return nbf_A_; }
+
+    /// Return number of atoms in the system
+    int get_natom_A() { return natom_A_; }
+
+    /// Return AO overlap S_nn
```
Comment (fevangelista) [inline src/orbital-helpers/fragment_projector.h]:
```suggestion
    /// Return the AO overlap matrix (S_nn)
```

what does `_nn` stand for?

File: src/orbital-helpers/pao_builder.cc [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ *
+ * @END LICENSE
+ */
+#include <map>
+#include <numeric>
+#include <regex>
```
[the remark this answers is not in the document]
Comment (fevangelista) [reply src/orbital-helpers/pao_builder.cc]: Can you check if the other includes are needed?

File: src/orbital-helpers/pao_builder.h [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+#include "psi4/libmints/basisset.h"
+#include "psi4/liboptions/liboptions.h"
+#include "psi4/libmints/dimension.h"
+
+#define _DEBUG_PAO_BUILDER_ 0
+
+namespace forte {
+
+/**
+ * @brief The fragment_projector class
```
Comment (fevangelista) [inline src/orbital-helpers/pao_builder.h]: This class description is incorrect, please update it.

File: src/orbital-helpers/pao_builder.h [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+ *
+ * A class to store information about a (system) fragment
+ */
+
+class PAObuilder {
+  public:
+    // ==> Constructors <==
```
Comment (fevangelista) [inline src/orbital-helpers/pao_builder.h]:
```suggestion
    // ==> Constructor <==
```

File: src/orbital-helpers/pao_builder.h [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+    PAObuilder(psi::SharedMatrix C, psi::Dimension noccpi, std::shared_ptr<psi::BasisSet> basis);
```
Comment (fevangelista) [inline src/orbital-helpers/pao_builder.h]: Can you briefly add a description of what PAOs do. I think other people in the future might want to use this class too.

File: tests/methods/embedding-3/input.dat [hunk 10]
```diff
@@ -1,7 +1,8 @@
 import forte
 
 refscf =   -115.6554016101969893
-refdsrgpt2 = -114.704305626419853
+#refdsrgpt2 = -114.704305626419853
```
Comment (fevangelista) [inline tests/methods/embedding-3/input.dat]: Remove comment here and the bit `# New value, should be lower than SCF energy`

File: pymodule.py (hunk 3 above)
Comment (henankf223) [reply pymodule.py]: Fixed

File: register_forte_options.py (hunk 4 above)
Comment (henankf223) [reply register_forte_options.py]: I will use ASET instead.

File: src/orbital-helpers/fragment_projector.h [hunk 11]
```diff
@@ -56,8 +56,16 @@ class FragmentProjector {
     // std::shared_ptr<BasisSet> minao_basis, 	std::shared_ptr<BasisSet> prime_basis);
 
     // Build projector and return AO basis matrix Pf_AO
-    psi::SharedMatrix build_f_projector(std::shared_ptr<psi::Molecule> molecule,
-                                        std::shared_ptr<psi::BasisSet> basis);
+	psi::SharedMatrix build_f_projector(std::shared_ptr<psi::BasisSet> basis);
```
[the remark this answers is not in the document]
Comment (henankf223) [reply src/orbital-helpers/fragment_projector.h]: Formatting updated.

File: src/orbital-helpers/fragment_projector.h [hunk 12]
```diff
@@ -56,8 +56,16 @@ class FragmentProjector {
...
+
+    /// Return number of basis functions in the system
```
[the remark this answers is not in the document]
Comment (henankf223) [reply src/orbital-helpers/fragment_projector.h]: I modify it to "Return number of system (fragment) basis functions"

File: src/orbital-helpers/fragment_projector.h (hunk 5 above)
Comment (henankf223) [reply src/orbital-helpers/fragment_projector.h]: It stands for a "square matrix". I think it is not necessary and I will remove this note.

File: src/orbital-helpers/pao_builder.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (henankf223) [reply src/orbital-helpers/pao_builder.cc]: Sure, I have checked and remove 4 includes.

File: src/orbital-helpers/pao_builder.h (hunk 7 above)
Comment (henankf223) [reply src/orbital-helpers/pao_builder.h]: I have revised the class descriptions in new commits.

File: src/orbital-helpers/pao_builder.h (hunk 9 above)
Comment (henankf223) [reply src/orbital-helpers/pao_builder.h]: I wrote an step-by-step explanation and a "typical usage" example

File: tests/methods/embedding-3/input.dat (hunk 10 above)
Comment (henankf223) [reply tests/methods/embedding-3/input.dat]: Removed

**Comment by fevangelista** [comment]
@henankf223: once the tests run we'll merge this in.

**Pull request merged**